### PR TITLE
Disable firestore tests by default

### PR DESCRIPTION
**Pull request by tcsc**
This PR originally intended to improve the reliability of the FirestoreDB unit tests. These tests run against a local fierestore emulator, and after much wailing and gnashing of teeth, I have come to the conclusion that the emulator is not a good match for what we need it to do.

The emulator  performs _very_erratically, sometimes missing snapshots, sometimes blocking on writes, and just generally being unpredictable - which is not good for an automated test system. 

Given that the tests are currently providing negative value I have disabled their execution. They can be enabled by defining the `TELEPORT_FIRESTORE_TEST` environment var (in the same manner as the `etcd` tests). This way, the tests will still be compiled so we can at least detect any obviously breaking changes.

Also; previously to this patch the DynamoDB tests were not compiled without the `dynamodb` build tag being set. To bring the DynamoDB back end into line with the others, the DynamoDB tests are now automatically compiled, but are skipped by default during a test run. They can be enabled by defining the `TELEPORT_DYNAMODB_TEST`

This PR still contains _some_ Firestore test cleanup and additional commentary from the original patch, but that is not the main point of the PR

### CODE REVIEW

**Comment by tcsc** [comment]
We're either doing something very wrong with the emulator, or the emulator itself is not very good. In some cases, the test fail because the snapshot stream flatly refuses to give us any events, even before we start fanning them out back up to the watcher created by the tests.

File: lib/backend/test/suite.go [hunk 1]
```diff
@@ -434,33 +434,41 @@ func testKeepAlive(t *testing.T, newBackend Constructor) {
 	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
 	defer cancel()
 
+	// When I create a new watcher...
 	watcher, err := uut.NewWatcher(ctx, backend.Watch{Prefixes: [][]byte{prefix("")}})
 	require.NoError(t, err)
 	defer func() { require.NoError(t, watcher.Close()) }()
 
+	// ...expect that the event channel contains the original `init` message
+	// sent when the Firestore client was wet up.
```
Comment (tcsc) [inline lib/backend/test/suite.go:443]:
```suggestion
	// sent when the Firestore client was set up.
```

**Review by codingllama — APPROVED** [review]
LGTM.

Looks like a more polished version of what I was trying the other day (except for my experiments with limit_watch_query=true). As you also mentioned, my experience with it was that while the individual KeepAlive test got better, the suite itself was still painfully flaky.

In hindsight, the fact that our backend test suite is tied to fake clocks likely wasn't the best choice, as it is prone to problems with backends that don't play nicely with having their clocks tweaked. I wonder if a rewrite to testKeepAlive (and testEvent?) to make them non-clock based could yield better results.

> We're either doing something very wrong with the emulator, or the emulator itself is not very good

I would start assuming the former - the emulator must be widely used and surely not every test suite that uses it flakes madly.

**Comment by codingllama** [comment]
FYI @zmb3

File: lib/backend/firestore/firestorebk.go [hunk 2]
```diff
@@ -293,6 +310,18 @@ func New(ctx context.Context, params backend.Params) (*Backend, error) {
...
 		go RetryingAsyncFunctionRunner(b.clientContext, linearConfig, b.Logger, b.purgeExpiredDocuments, "purgeExpiredDocuments")
 	}
+
+	l.Debug("Waiting for event watcher to start")
+	select {
+	case <-ctx.Done():
+		l.Error("Timed out waiting for event monitor")
+		return nil, ctx.Err()
+
+	case <-b.watchStarted.Done():
```
Comment (zmb3) [inline lib/backend/firestore/firestorebk.go:320]: I think the `watchStarted` context is something that was recently deprecated, and @fspmarshall was planning to remove from all of the backends. We didn't want `New` to block, and instead wanted watchers to rely on the `OpInit` event to know when the watch is ready.

With this change, the firestore backend would be the only one using this context, even though it still exists in the code for the other backend types.

I think we ought to remain consistent and either use this approach for all backends or for no backends.

Comment (tcsc) [reply lib/backend/firestore/firestorebk.go:320]: OK, will revert

**Comment by tcsc** [comment]
> I would start assuming the former - the emulator must be widely used and surely not every test suite that uses it flakes madly.

Also, if we just assume that the emulator is broken then our tests will never get any better.

File: lib/backend/firestore/firestorebk_test.go [hunk 3]
```diff
@@ -52,15 +53,25 @@ func TestMarshal(t *testing.T) {
...
-	"collection_name":                   "tp-cluster-data-test",
-	"project_id":                        "tp-testproj",
-	"endpoint":                          "localhost:8618",
-	"purgeExpiredDocumentsPollInterval": time.Second,
+// var commonFirestoreParams = map[string]interface{}{
+// 	"collection_name":                   "tp-cluster-data-test",
+// 	"project_id":                        "tp-testproj",
+// 	"endpoint":                          "localhost:8618",
+// 	"purgeExpiredDocumentsPollInterval": time.Second,
+// }
```
Comment (codingllama) [inline lib/backend/firestore/firestorebk_test.go:61]: Delete dead code?

File: lib/backend/firestore/firestorebk_test.go [hunk 4]
```diff
@@ -52,15 +53,25 @@ func TestMarshal(t *testing.T) {
...
+
+func firestoreParams(t *testing.T) map[string]interface{} {
```
Comment (codingllama) [inline lib/backend/firestore/firestorebk_test.go:63]: Maybe:

```suggestion
type firestoreConfig map[string]interface

func newFirestoreConfig(t *testing.T) firestoreConfig {
```

and use `firestoreConfig` throughout?

**Review by codingllama — APPROVED** [review]
LGTM.

Did you notice any difference in flakiness after removing the watch context?

File: lib/backend/firestore/firestorebk.go (hunk 2 above)
Comment (tcsc) [reply lib/backend/firestore/firestorebk.go]: Deleted entirely

**Comment by tcsc** [comment]
OK, I admit defeat. I have re-purposed this to disable the Firestore tests by default.

File: lib/backend/dynamo/dynamodbbk_test.go [hunk 5]
```diff
@@ -38,7 +36,16 @@ func TestMain(m *testing.M) {
 
 const tableName = "teleport.dynamo.test"
 
+func ensureTestsEnabled(t *testing.T) {
+	const varName = "TELEPORT_DYNAMODB_TEST"
+	if os.Getenv(varName) == "" {
+		t.Skip("Firestore tests are disabled. Enable by defining the " + varName + " environment variable")
```
Comment (codingllama) [inline lib/backend/dynamo/dynamodbbk_test.go:42]:
```suggestion
		t.Skipf("DynamoDB tests are disabled. Enable by defining the %v environment variable", varName)
```

File: lib/backend/firestore/firestorebk_test.go [hunk 6]
```diff
@@ -52,15 +52,38 @@ func TestMarshal(t *testing.T) {
...
+		"project_id":                        "tp-testproj",
+		"endpoint":                          "localhost:8618",
+		"purgeExpiredDocumentsPollInterval": time.Second,
+	}
+}
+
+func ensureTestsEnabled(t *testing.T) {
+	const varName = "TELEPORT_FIRESTORE_TEST"
+	if os.Getenv(varName) == "" {
+		t.Skip("Firestore tests are disabled. Enable by defining the " + varName + " environment variable")
```
Comment (codingllama) [inline lib/backend/firestore/firestorebk_test.go:71]:
```suggestion
		t.Skipf("Firestore tests are disabled. Enable by defining the %v environment variable", varName)
```

**Review by codingllama — APPROVED** [review]
LGTM.

Thanks for the effort you put into this, it was definitely a difficult one.

File: lib/backend/firestore/firestorebk_test.go [hunk 7]
```diff
@@ -52,15 +52,38 @@ func TestMarshal(t *testing.T) {
 	assert.NoError(t, err)
 }
 
-var commonFirestoreParams = map[string]interface{}{
-	"collection_name":                   "tp-cluster-data-test",
-	"project_id":                        "tp-testproj",
-	"endpoint":                          "localhost:8618",
-	"purgeExpiredDocumentsPollInterval": time.Second,
+func firestoreParams(t *testing.T) backend.Params {
```
Comment (zmb3) [inline lib/backend/firestore/firestorebk_test.go:55]: Why do you need `t` here?

Comment (tcsc) [reply lib/backend/firestore/firestorebk_test.go:55]: Inertia. Removed.

**Comment by russjones** [comment]
@nklaassen Can you take a look, looks like this is the last thing we need to stabilize unit tests on master.

File: lib/backend/test/suite.go [hunk 8]
```diff
@@ -578,7 +587,7 @@ func requireEvent(t *testing.T, watcher backend.Watcher, eventType types.OpType,
 		require.FailNow(t, "Watcher has unexpectedly closed.")
 
 	case <-time.After(timeout):
-		require.FailNow(t, "Timed out after %v waiting for event %v", timeout, eventType.String())
+		require.FailNowf(t, "Timed out", "Timed out after %v waiting for event %v", timeout.String(), eventType.String())
```
Comment (nklaassen) [inline lib/backend/test/suite.go:590]: nit:
```suggestion
		require.FailNowf(t, "Timed out after %v waiting for event %v", timeout.String(), eventType.String())
```

**Review by nklaassen — APPROVED** [review]

File: lib/backend/test/suite.go (hunk 8 above)
Comment (tcsc) [reply lib/backend/test/suite.go:590]: That will give errors like:
```
    firestorebk_test.go:158: 
                Error Trace:    firestorebk_test.go:158
                Error:          Timed out after %v waiting for event %v
                Test:           TestFail
                Messages:       10s%!(EXTRA string=some event)
--- FAIL: TestFail (0.00s)
```

My change makes it give:
```
    firestorebk_test.go:158: 
                Error Trace:    firestorebk_test.go:158
                Error:          Timed out
                Test:           TestFail
                Messages:       Timed out after 10s waiting for event some event
--- FAIL: TestFail (0.00s)
```

File: lib/backend/firestore/firestorebk.go [hunk 9]
```diff
@@ -238,17 +235,28 @@ func CreateFirestoreClients(ctx context.Context, projectID string, endPoint stri
...
+// construction
+type Option func(*Backend)
+
+// WithClock provides a custom clock the firestore Backend
+func WithClock(clock clockwork.Clock) Option {
+	return func(b *Backend) {
+		b.clock = clock
+	}
+}
+
```
Comment (russjones) [inline lib/backend/firestore/firestorebk.go:248]: If you are making this change, just follow passing a config struct approach we're doing now. You might want to look at how DynamoDB events backend is now.

**Review by russjones — APPROVED** [review]

File: lib/backend/firestore/firestorebk.go (hunk 9 above)
Comment (tcsc) [reply lib/backend/firestore/firestorebk.go]: Done. `Config` name was already taken, so it's named `Options` instead.

**Pull request merged**